cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java

```diff
@@ -335,24 +345,55 @@ protected boolean canJoinOnSQLEngine(String stageName,
     return !containsBroadcastStage;
   }
 
+  /**
+   * Check if this stage is configured as a stage that should be pushed to the SQL engine
+   * @param stageName stage name
+   * @return boolean stating if this stage should always try to be executed in the SQL engine.
+   */
+  protected boolean shouldForcePushToSQLEngine(String stageName) {
+    return sqlEngineAdapter != null && sqlEngineAdapter.getIncludedStageNames().contains(stageName);
+  }
+
+  /**
+   * Check if this stage is configured as a stage that should never be pushed to the SQL engine
+   * @param stageName stage name
+   * @return boolean stating if this stage should always be skipped from executing in the SQL engine.
+   */
+  protected boolean shouldForceSkipSQLEngine(String stageName) {
+    return sqlEngineAdapter != null && sqlEngineAdapter.getExcludedStageNames().contains(stageName);
+  }
+
   /**
    * If SQL Engine is present, supports relational transform and current stage data is already
    * provided by SQL engine, adds SQL Engine implementation of relational engine
    * @param stageData
    * @return
    */
   @Override
-  protected Iterable<SparkCollectionRelationalEngine> getRelationalEngines(SparkCollection<Object> stageData) {
-    if (sqlEngineAdapter == null || !sqlEngineAdapter.supportsRelationalTranform()
-      || !(stageData instanceof SQLBackedCollection)) {
+  protected Iterable<SparkCollectionRelationalEngine> getRelationalEngines(String stageName,
+                                                                           StageSpec stageSpec,
+                                                                           SparkCollection<Object> stageData) {
+    if (sqlEngineAdapter == null || !sqlEngineAdapter.supportsRelationalTranform()) {
       //Relational transform on SQL engine is not supported
-      return super.getRelationalEngines(stageData);
+      return super.getRelationalEngines(stageName, stageSpec, stageData);
     }
+
+    // Explicitly skip this stage if the stage is configured as an excluded stage.
+    if (shouldForceSkipSQLEngine(stageName)) {
+      return super.getRelationalEngines(stageName, stageSpec, stageData);
+    }
+
+    // If this stage is not pushed down and it's not in the included stages, we can skip relational transformation
+    // in the SQL engine.
+    if (!(stageData instanceof SQLBackedCollection || shouldForcePushToSQLEngine(stageName))) {
```

Comment:
nit: I am wondering if it would be more readable as `!A && !B` vs `!(A || B)`. It would more directly correspond to the comment.

Reply:
Updated